File: src/HardEngine.hs
module HardEngine (hardEngine) where 
import Data.Bits
import RandomEngine

hardEngine :: [Int] -> (Int,Int)
hardEngine list = do
    let xor4 = getXorMod4 list;
    if xor4 == 0 then randomEngine 0 list
    else makeBestMove xor4 list 0

getXorMod4 :: [Int] -> Int
getXorMod4 [] = 0
getXorMod4 (x:xs) = xor (rem x 4) (getXorMod4 xs)


makeBestMove :: Int -> [Int] -> Int -> (Int,Int)
makeBestMove xOr (x:xs) i =
    if x-1 >= 0 && xor (xor xOr x) (x-1) == 0 then (1, i)
    else if x-2 >= 0 && xor (xor xOr x) (x-2) == 0 then (2, i)
    else if x-3 >= 0 && xor (xor xOr x) (x-3) == 0 then (3, i)
    else makeBestMove xOr xs (i+1)
    

File: prolog/src/Engines/HardEngine.pl
:- include('RandomEngine.pl').

xorMod4([], B, B).
xorMod4([H | T], B, X) :-
	UB is B xor (H mod 4),
	xorMod4(T, UB, X).

move([H | T], [V, Idx], I, X) :-
    (
        H-1 >= 0, X xor H xor (H-1) =:= 0, V is 1, Idx is I;
        H-2 >= 0, X xor H xor (H-2) =:= 0, V is 2, Idx is I;
        H-3 >= 0, X xor H xor (H-3) =:= 0, V is 3, Idx is I
    );
    move(T, [V, Idx], I+1, X).

hardEngine([H | T], [V, Idx]) :-
    xorMod4([H | T], 0, 0),
    randomEngine([H | T], [V, Idx]),
    !.

hardEngine([H | T], [V, Idx]) :-
    xorMod4([H | T], 0, X),
    X \= 0,
    move([H | T], [V, Idx], 0, X),
    !.
